File: ex2.py
#!/usr/bin/python3

import sys
import math

def dijkstra(adj, n, m, v):
    costs = [math.inf] * n
    costs[v] = 0

    v_open = []
    for i in range(0, n):
        v_open.append(i)

    while len(v_open) != 0:
        closest = v_open[0]
        for o in v_open:
            closest = costs[o] < costs[closest] if o else closest
        if closest in v_open:
            v_open.remove(closest)
        for o in v_open:
            if adj[closest * n + o] != 0:
                costs[o] = min([costs[o], costs[closest] + adj[closest * n + o]])

    return costs

def solve(adj, n, m, k, c):
    costs = dijkstra(adj, n, m, k)
    return costs[c - 1]

n = 0
m = 0
k = 0
c = 0

while (True):

    line = sys.stdin.readline().split()
    n = int(line[0])
    m = int(line[1])
    c = int(line[2])
    k = int(line[3])

    if n == 0 and m == 0 and c == 0 and k == 0:
        break

    adj = [0] * (n * n)

    for i in range(0, m):

        u = 0
        v = 0
        t = 0

        edge = sys.stdin.readline().split()
        u = int(edge[0])
        v = int(edge[1])
        t = int(edge[2])

        adj[u * n + v] = t
        adj[v * n + u] = t

        print(solve(adj, n, m, c, k))

File: ex2.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <alloca.h>

#define MIN(i, j) (i < j) ? i : j;
#define INFINITY INT_MAX / 2

typedef struct graph
{
	int n, m;
	int *adj;
} Graph;

#define SET(g, i, j, c) g.adj[i * g.n + j] = c
#define GET(g, i, j) g.adj[i * g.n + j]

Graph CreateGraph(int n, int m);
void DestroyGraph(Graph g);
int Solve(Graph g, int k, int c);

int main(void)
{
	int n, m, c, k;

	while (1)
	{
		int scanRes = scanf("%d %d %d %d\n", &n, &m, &c, &k);

		if (n == 0 && m == 0 && c == 0 && k == 0) break;

		Graph g = CreateGraph(n, m);

		int i;
		for (i = 0; i < m; ++i)
		{
			int u, v, c;
			scanRes = scanf("%d %d %d\n", &u, &v, &c);

			SET(g, u, v, c);
			SET(g, v, u, c);
		}

		int j;
		for (i = 0; i < c; ++i)
			for (j = 0; j < n; ++j)
				if (j != i + 1)
					SET(g, i, j, 0);

		printf("%d\n", Solve(g, k, c));

		DestroyGraph(g);
	}

	return 0;
}

Graph CreateGraph(int n, int m)
{
	Graph g;

	g.n = n;
	g.m = m;

	g.adj = malloc(sizeof(int) * n * n);

	for (int i = 0; i < n * n; ++i)
		g.adj[i] = 0;

	return g;
}

void DestroyGraph(Graph g)
{
	free(g.adj);
}

int Solve(Graph g, int k, int c)
{
	int result = 0;

	/* Executing Dijkstra */ {

		int *costs = alloca(sizeof(int) * g.n);
		int *open = alloca(sizeof(int) * g.n);
		int size = g.n;

		int count;
		for (count = 0; count < g.n; ++count)
		{
			open[count] = count;
			if (count == k) costs[count] = 0;
			else costs[count] = INFINITY;
		}

		while (size > 0)
		{
			// Getting closest open node to start node:
			int closest = open[0];
			int indexOpen;
			for (indexOpen = 0; indexOpen < size; ++indexOpen)
				closest = (costs[indexOpen] < costs[closest]) ? indexOpen : closest;

			// Closing the closest node:
			int remove;
			for (remove = indexOpen; remove < size - 1; ++remove)
				open[remove] = open[remove + 1];
			--size;

			for (indexOpen = 0; indexOpen < size; ++indexOpen)
				if (GET(g, closest, indexOpen))
					costs[indexOpen] = MIN(costs[indexOpen], costs[closest] + GET(g, closest, indexOpen));
		}

		result = costs[c - 1];
	}

	return result;
}
